ToDo REST API 만들기

csh9066 opened the pull request:
e2e 테스트와 api 테스트 모두 통과 하도록 했습니다.
요번 주도 잘 부탁드립니다!

csh9066 committed:
스프링 MVC을 이용해 테스트를 통과하는 TaskController를 만들어라

csh9066 committed:
updateTask에 잘못 표기한 변수명들을 수정해라

기존 코드에서는 target(수정해야 할 task) source(수정 요청을 한 task)가 반대로 되어있어서 수정했습니다.

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/repository/TaskRepository.java:
```diff
@@ -0,0 +1,37 @@
+    }
+
+    public Task save(Task task) {
+        task.setId(sequence);
```
`setId` 를 한번 없애보면 좋을 것 같아요!

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/models/Task.java:
```diff
@@ -0,0 +1,28 @@
+package com.codesoom.assignment.models;
+
+public class Task {
```
`setId` `setTitle` 을 없애보는 시도를 해보면 좋은 경험이 될 거에요!

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/repository/TaskRepository.java:
```diff
@@ -0,0 +1,37 @@
+
+    private void incrementSequence() {
+        this.sequence += 1;
+    }
```
한번 병렬성을 고민해서 개선해보도록 해요!

csh9066 committed:
Task의 setter 대신 좀 더 명확한 이름의 메소드로 변경해라

csh9066 committed:
TaskRepository를 thread safety하게 변경해라

csh9066 committed:
save 메소드의 변수명을 명확하게 변경해라

새로운 Task를 만드는데 target이란 변수명은 안 어울리는거 같다 newTask로 좀 더 명확하게 변경했다.

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/repository/TaskRepository.java:
```diff
@@ -0,0 +1,40 @@
+        Task newTask = new Task(sequence.get(), source.getTitle());
+        taskMap.put(newTask.getId(), newTask);
+
+        incrementSequence();
```
좋은시도입니다! 하지만 한가지 아쉬운 점이 있어요!
그것은 아래와 같은 코드를 보고 돌려보면 알 수 있습니다. 예시를 한번 간단하게 만들어보았어요! 한번 실행해보고 개선해보도록해요!
```java
    public static void main(String[] args) throws InterruptedException {
        final List<Thread> threadList = new ArrayList<>();
        final TaskRepository taskRepository = new TaskRepository();
        for (int i = 0; i < 1000; i++) {
            threadList.add(new Thread(() -> {
                try {
                    Thread.sleep(100);
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
                taskRepository.save(new Task(0L, "TITLE-"));
            }));
        }

        for (Thread r : threadList) r.start();
        for (Thread r : threadList) r.join();

        final List<Task> tasks = taskRepository.findAll();
        System.out.println(tasks.size());
        System.out.println(tasks);
    }
```

csh9066 replied:
감사합니다! 덕분에 여러가지 방법으로 테스트 할 수 있었어요
사실 동시성에 대해 지식이 전무했는데 요번 기회를 통해 찍어는 먹어본거 같아요
하지만 아직 갈길이 먼거같아요ㅜㅜ

csh9066 committed:
TaskRepository의 save시 AtomicLong을 올바르게 사용해라

private 메소드인 incrementSequence가 AtomicLong을 올바르게 사용하지 않아 문제를 일으켰습니다 save를 할때 AtomicLong의 getAndIncrement를 이용해 올바르게 save가 작동하도록 변경했습니다.

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/repository/TaskRepository.java:
```diff
@@ -21,20 +24,12 @@ public List<Task> findAll() {
 
     public Task save(Task source) {
-        Task newTask = new Task(sequence.get(), source.getTitle());
+        Task newTask = new Task(sequence.getAndIncrement(), source.getTitle());
```
너무 좋아요! 기억하세요! 조회한 값으로 수정을 한다면 조회와 수정이 하나의 트랜잭션으로 일어나야 합니다!

csh9066 replied:
넵 명심하겠습니다!

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/controllers/TaskController.java:
```diff
@@ -0,0 +1,52 @@
+@CrossOrigin
+@RequestMapping("/tasks")
+@RestController
+public class TaskController {
```
Controller에서 직접적으로 Repository를 바라보고 있군요!

당연히 이렇게 간단한 Crud에서는 문제가 없습니다.

만약에 `Task`가 아니라 결재를 하는 거라고 우리는 상상해봐요! 좀 더 도메인이 복잡하겠죠! 

Controller는 유저의 요청에 따라 그것을 해결할 수 있는 객체에게 위임하는 것만으로 벅차다고 생각해봐요! 

결론 ! : `TaskController` 에서 `TaskRepository`를 없애보세요!

csh9066 committed:
Task의 비지니스 로직을 처리하는 TaskSercice를 구현하라

csh9066 committed:
TaskController는 TaskService에게 비지니스 로직을 위임하도록 변경해라

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/service/TaskService.java:
```diff
@@ -0,0 +1,46 @@
+import java.util.List;
+
+@Service
+public class TaskService {
```
이 서비스를 쪼개볼 수 있을까요?! 

미래에는 벌써 할일에 대한 추가기획사항이 몰려오고 있다고 상상해봐요! 
`TaskService` 를 쪼개지 않으면 수천줄의 코드를 가진 레거시로서 위상을 떨치기 시작할거에요!

우리는 이것들을 미리 쪼개보는거에요!

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/controllers/TaskController.java:
```diff
@@ -0,0 +1,48 @@
+
+    @ResponseStatus(HttpStatus.CREATED)
+    @PostMapping
+    public Task createTask(@RequestBody Task task) {
```
할일은 `id` 가 없이도 할일이라고 할 수 있을까요?!

할일과 할일생성요청에 대한 객체를 분리해보는거에요!

csh9066 committed:
TaskService의 책임을 분리하여라

csh9066 committed:
컨트롤러에서 Task 도메인을 보호하는 Dto 객체를 만들어라

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/service/TaskFinishService.java:
```diff
@@ -0,0 +1,20 @@
+import org.springframework.stereotype.Service;
+
+@Service
+public class TaskFinishService {
```
왠지 할일을 완료시키는 서비스 같군요! 

하지만 주석이 있으면 좋겠어요!

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/service/TaskUpdateService.java:
```diff
@@ -0,0 +1,23 @@
+        this.taskRepository = taskRepository;
+    }
+
+    public Task update(Long id, Task source) {
```
`Task` 안에 `id` 가 null인 상태라면 `Task` 말고 title만 받는 것도 방법이에요! 

혹은 업데이트만을 위한 객체를 만들어서 쓰는거에요!

예를들면 이렇게요!

```java
public Task update(TaskUpdateCommand taskUpdateCommand) {
  Task target = taskRepository.findById(taskUpdateCommand.getId())
                .orElseThrow(TaskNotFoundException::new);
  target.updateTitle(taskUpdateCommand.getTitle());
}
```

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/repository/TaskRepository.java:
```diff
@@ -0,0 +1,35 @@
+        return new ArrayList<>(taskMap.values());
+    }
+
+    public Task save(Task source) {
```
`Task` 가 `id` 가 null 일 때는 `Task` 인가를 고민해보는 것이 좋을 것 같아요!

지금은 `JPA` 를 쓰는 것이 아니니까! 이런 고민을 해보는 것이 도움이 될 것 같아요! 

만약에 `Task` 를 생성하려면 `id` 가 없어도 된다고 결론이 난다면, 주석에 그부분을 명시해 놓는 것이 좋아요!

csh9066 replied:
좋은 리뷰 감사합니다! 
뭔가 찝찝하다고 생각한 부분에대해 다 짚어주셔서 놀랐어요!
한 주 동안 계속 도전 과제를 주셔서 여러 가지로 많이 생각할 수 있어서 좋았습니다!

csh9066 committed:
TaskRepository의 save 메소드의 매개변수를 변경해라

csh9066 committed:
Task의 update 메소드의 이름을 더 명확하게 변경해라

csh9066 committed:
TaskFinishService의 이름을 좀 더 명확하게 변경해라

ssisksl77 commented on app/src/test/java/com/codesoom/assignment/TaskRepositorySyncTest.java:
```diff
@@ -26,7 +26,7 @@ void save() throws InterruptedException {
                     e.printStackTrace();
                 }
-                taskRepository.save(new Task(0L, "TITLE"));
+                taskRepository.save("Title");
```
좋아요! 어쩌면 이런 생각을 하실거에요!
"아니 문자열이 그럼 2,3개가 되면 순서가 복잡해지면서 실수가 많아질 것 같은데...?!"

그러므로 나중에는 이 문자열마저 객체화 시키면 되요! `Title` 이라는 객체를 만들면 되겠죠!

csh9066 replied:
영환님 궁금한게 있습니다!
`Title` 이라는 객체를 만들어 봤는데요 여기서 궁금증이 있습니다
단순히 Task를 생성할 때 문자열이 많아서 실수할 거 같으면 빌더 패턴같은게 더 괜찮지 않나요?

csh9066 committed:
Task 안전하게 생성할 수 있게 Title 을 구현해라
